### PR TITLE
feat: crl caching

### PR DESCRIPTION
**Pull request by rjzak**
Refs https://github.com/enarx/enarx/issues/2389

Signed-off-by: Richard Zak <richard@profian.com>

### CODE REVIEW

File: src/cli/platform/snp/crl.rs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+use crate::backend::sev::snp::vcek::sev_cache_dir;
+
+use std::process::ExitCode;
+
+use clap::Args;
+
+const GENOA: &str = "https://kdsintf.amd.com/vcek/v1/Genoa/crl";
+const MILAN: &str = "https://kdsintf.amd.com/vcek/v1/Milan/crl";
+
+/// Fetch Intel's Certificate Revocation Lists (CRLs),
```
Comment (haraldh) [inline src/cli/platform/snp/crl.rs:13]: Intel?

File: src/cli/platform/snp/crl.rs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+/// saving as cached files in `/var/cache/intel-sgx/` directory
```
Comment (haraldh) [inline src/cli/platform/snp/crl.rs:14]: sgx?

**Review by haraldh — CHANGES_REQUESTED** [review]

File: src/cli/platform/snp/crl.rs (hunk 2 above)
Comment (rjzak) [reply src/cli/platform/snp/crl.rs]: Ooops. Fixed.

File: src/cli/platform/snp/crl.rs (hunk 1 above)
Comment (rjzak) [reply src/cli/platform/snp/crl.rs]: Fixed.

File: src/cli/platform/caching.rs [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+use std::fs::OpenOptions;
+use std::io::{Read, Write};
+use std::path::Path;
+
+use sha2::{Digest, Sha256};
+
+/// Fetch a URL and save the contents as the hash of the URL
+pub fn save_file(url: &str, dest: &Path) -> anyhow::Result<()> {
+    let response = ureq::get(url).call()?.into_reader();
+    let response: Vec<u8> = response.bytes().map(|d| d.unwrap()).collect();
```
Comment (haraldh) [inline src/cli/platform/caching.rs:12]: Scratch this, and use std::io:: copy() instead of write_all below.

File: src/cli/platform/caching.rs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    let mut dest = dest.to_path_buf();
+    dest.push(hex::encode(Sha256::digest(url)));
+
+    let mut file = OpenOptions::new()
+        .create(true)
+        .write(true)
+        .open(dest)
+        .unwrap();
+
+    file.write_all(&response)?;
```
Comment (haraldh) [inline src/cli/platform/caching.rs:23]: Use io::copy()

File: src/cli/platform/caching.rs [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    file.flush()?;
```
Comment (haraldh) [inline src/cli/platform/caching.rs:24]: Why flush? Is close() not enough?

File: src/cli/platform/caching.rs [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    let response: Vec<u8> = response.bytes().map(|d| d.unwrap()).collect();
+
+    let mut dest = dest.to_path_buf();
+    dest.push(hex::encode(Sha256::digest(url)));
+
+    let mut file = OpenOptions::new()
+        .create(true)
+        .write(true)
+        .open(dest)
+        .unwrap();
```
Comment (haraldh) [inline src/cli/platform/caching.rs:21]: Unwrap? Not a very nice error message.

File: src/cli/platform/caching.rs (hunk 4 above)
Comment (haraldh) [reply src/cli/platform/caching.rs:23]: And add a context()

File: src/cli/platform/caching.rs [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+use std::fs::OpenOptions;
+use std::io::{Read, Write};
+use std::path::Path;
+
+use sha2::{Digest, Sha256};
+
+/// Fetch a URL and save the contents as the hash of the URL
+pub fn save_file(url: &str, dest: &Path) -> anyhow::Result<()> {
+    let response = ureq::get(url).call()?.into_reader();
```
Comment (haraldh) [inline src/cli/platform/caching.rs:11]: Please add context() for all error returns.

**Review by haraldh — CHANGES_REQUESTED** [review]

File: src/cli/platform/caching.rs (hunk 3 above)
Comment (rjzak) [reply src/cli/platform/caching.rs:12]: How does the `ureq` response get turned into a reader for `copy()`? I'm getting an error that `std::io::Read` is not implemented for `Vec<u8>`, and `ureq::get(url).call()?.into_reader()` isn't helpful either.

Why is `copy()` desirable over `file.write_all()`?

**Comment by rjzak** [comment]
What does the CI job "Pull Request Labeler" want? I put a label on the PR, and it still has errors.

```
Run actions/labeler@v4
  with:
    repo-token: ***
    configuration-path: .github/labeler.yml
    sync-labels: false
Error: HttpError: Resource not accessible by integration
Error: Resource not accessible by integration
```

File: src/cli/platform/caching.rs (hunk 6 above)
Comment (rjzak) [reply src/cli/platform/caching.rs]: Removed, added `context()`

File: src/cli/platform/caching.rs (hunk 5 above)
Comment (rjzak) [reply src/cli/platform/caching.rs]: Removed, added `context()`

File: src/cli/platform/caching.rs (hunk 3 above)
Comment (haraldh) [reply src/cli/platform/caching.rs:16]: Don't use the Vec, use the into_reader() return value for io::copy

File: src/cli/platform/caching.rs [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    let response: Vec<u8> = response.bytes().map(|d| d.unwrap()).collect();
+
+    let mut dest = dest.to_path_buf();
+    dest.push(hex::encode(Sha256::digest(url)));
+
+    let mut file = OpenOptions::new()
+        .create(true)
+        .write(true)
+        .open(dest)
+        .context("opening destination file for saving CRL")?;
```
Comment (haraldh) [inline src/cli/platform/caching.rs:25]: Filename? Path?

File: src/cli/platform/caching.rs [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+use std::io::{Read, Write};
+use std::path::Path;
+
+use sha2::{Digest, Sha256};
+
+/// Fetch a URL and save the contents as the hash of the URL
+pub fn save_file(url: &str, dest: &Path) -> anyhow::Result<()> {
+    let response = ureq::get(url)
+        .call()
+        .context("retrieving CRL from server")?
```
Comment (haraldh) [inline src/cli/platform/caching.rs:14]: Which url?

**Review by haraldh — CHANGES_REQUESTED** [review]

File: src/cli/platform/caching.rs (hunk 3 above)
Comment (rjzak) [reply src/cli/platform/caching.rs:16]: @haraldh

File: src/cli/platform/caching.rs (hunk 8 above)
Comment (rjzak) [reply src/cli/platform/caching.rs]: Added

File: src/cli/platform/caching.rs (hunk 9 above)
Comment (rjzak) [reply src/cli/platform/caching.rs]: Added

File: src/cli/platform/caching.rs [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+            "opening destination file {:?} for saving CRL {url}",
+            dest.display()
+        ))?;
+
+    file.write_all(&response).context(format!(
+        "writing CRL file {} to disk from {url}",
+        dest.display()
+    ))?;
+    Ok(())
+}
```
Comment (haraldh) [inline src/cli/platform/caching.rs:35]:
```suggestion
pub fn save_file(url: &str, dest: &Path) -> anyhow::Result<()> {
    let mut response = ureq::get(url)
        .call()
        .context(format!("retrieving CRL {url} from server"))?
        .into_reader();

    let mut dest = dest.to_path_buf();
    dest.push(hex::encode(Sha256::digest(url)));

    let mut file = OpenOptions::new()
        .create(true)
        .write(true)
        .open(&dest)
        .context(format!(
            "opening destination file {:?} for saving CRL {url}",
            dest.display()
        ))?;

    std::io::copy(&mut response, &mut file).context(format!(
        "saving CRL {url} to destination file {:?}",
        dest.display()
    ))?;

    Ok(())
}
```

**Review by haraldh — CHANGES_REQUESTED** [review]

File: src/cli/platform/caching.rs (hunk 10 above)
Comment (rjzak) [reply src/cli/platform/caching.rs]: Thank you for that. I don't understand what I did wrong when I tried this, but I was getting errors about `Box<...>` not implementing some needed trait. Works now.

**Review by haraldh — APPROVED** [review]

**Pull request merged**